packages/stock-location-next/src/services/stock-location.ts

```diff
@@ -117,13 +121,55 @@ export default class StockLocationModuleService<
     return await this.stockLocationService_.create(data, context)
   }
 
+  async upsert(
+    data: UpsertStockLocationInput,
+    context?: Context
+  ): Promise<StockLocationTypes.StockLocationDTO>
+  async upsert(
+    data: UpsertStockLocationInput[],
+    context?: Context
+  ): Promise<StockLocationTypes.StockLocationDTO[]>
+
+  @InjectManager("baseRepository_")
+  async upsert(
+    data: UpsertStockLocationInput | UpsertStockLocationInput[],
+    @MedusaContext() context: Context = {}
+  ): Promise<
+    StockLocationTypes.StockLocationDTO | StockLocationTypes.StockLocationDTO[]
+  > {
+    const input = Array.isArray(data) ? data : [data]
+    const toUpdate = input.filter(
+      (location): location is UpdateStockLocationNextInput => !!location.id
+    ) as UpdateStockLocationNextInput[]
+    const toCreate = input.filter(
+      (location) => !location.id
+    ) as CreateStockLocationInput[]
+
+    const operations: Promise<StockLocation[]>[] = []
+
+    if (toCreate.length) {
+      operations.push(this.create_(toCreate, context))
+    }
+    if (toUpdate.length) {
+      operations.push(this.update_(toUpdate, context))
+    }
```

Comment:
todo: those two operations happen in different transactions, should we create an upsert_ method that is wrapped in a single transaction to prevent any unexpected side effects?